orangecontrib/prototypes/widgets/owinteractions.py

```diff
@@ -82,27 +107,108 @@ def get_states(self, initial_state):
 		return states
 
 
+class InteractionItemDelegate(gui.TableBarItem):
+	def __init__(self, **kwargs):
+		super().__init__(**kwargs)
+		self.r = QColor(255, 170, 127)
+		self.g = QColor(170, 242, 43)
+		self.b = QColor(70, 190, 250)
+		self.__line = QLineF()
+		self.__pen = QPen(self.b, 5, Qt.SolidLine, Qt.RoundCap)
+
+	def paint(
+			self, painter: QPainter, option: QStyleOptionViewItem,
+			index: QModelIndex
+	) -> None:
+		opt = QStyleOptionViewItem(option)
+		self.initStyleOption(opt, index)
+		widget = option.widget
+		style = QApplication.style() if widget is None else widget.style()
+		self.__style = style
+		text = opt.text
+		opt.text = ""
+		style.drawControl(QStyle.CE_ItemViewItem, opt, painter, widget)
+		textrect = style.subElementRect(
+			QStyle.SE_ItemViewItemText, opt, widget)
+
+		center = self.cachedData(index, InteractionRank.IntRole)
+		if center is not None:
+			left, right = self.cachedData(index, InteractionRank.GainRole)
+			left, right = abs(left), abs(right)
+			color = self.g if center > 0 else self.r
+			center = abs(center)
+
+			rect = option.rect
+			pw = self.penWidth
+			hmargin = 3 + pw / 2  # + half pen width for the round line cap
+			vmargin = 1
+			textoffset = pw + vmargin * 2
+			baseline = rect.bottom() - textoffset / 2
+			width = (rect.width() - 2 * hmargin) / 3
+			painter.save()
+			painter.setRenderHint(QPainter.Antialiasing)
+			pen = self.__pen
+			pen.setColor(self.b)
+			pen.setWidth(pw)
+			painter.setPen(pen)
+			line = self.__line
+			line.setLine(
+				rect.left() + hmargin,
+				baseline,
+				rect.left() + hmargin + width * left,
+				baseline
+			)
+			painter.drawLine(line)
+			line.setLine(
+				rect.left() + 3*hmargin + width * (left + center),
+				baseline,
+				rect.left() + 3*hmargin + width * (left + center + right),
+				baseline
+			)
+			painter.drawLine(line)
+			pen.setColor(color)
+			painter.setPen(pen)
+			line.setLine(
+				rect.left() + 2*hmargin + width * left,
+				baseline,
+				rect.left() + 2*hmargin + width * (left + center),
+				baseline
+			)
+			painter.drawLine(line)
```

Comment:
These setline/drawline code pattern repeats a lo. They should be probably converted to an internal function which will be then called three times.